### PR TITLE
OC-141: Add support for text blocks

### LINKED ISSUE #141: Java 15 Text blocks not supported
I'm receiving an error running openclover on my code base.

Java: 15.0.1
openclover maven plugin 4.4.1
maven: 3.6.3
OS: macos big SUR 11.0.1

code on line 5 of SqlConst:
```
    public static final String INSERT_PLAYLIST_IDENTIFICTION = """
            INSERT INTO PLAYLIST_ENTRY(
                 playlist, line_number, play_time, played_on, isrc)
            VALUES(?, ?, ?, ?, ?)
            """;
```

stacktrace:
```
SqlConst.java:5:67:expecting '"', found '\n'
    at com.atlassian.clover.instr.java.Instrumenter.instrument (Instrumenter.java:164)
    at com.atlassian.clover.CloverInstr.execute (CloverInstr.java:110)
    at com.atlassian.clover.CloverInstr.mainImpl (CloverInstr.java:86)
    at com.atlassian.maven.plugin.clover.internal.instrumentation.AbstractInstrumenter.instrumentSources (AbstractInstrumenter.java:205)
    at com.atlassian.maven.plugin.clover.internal.instrumentation.AbstractInstrumenter.instrument (AbstractInstrumenter.java:80)
    at com.atlassian.maven.plugin.clover.CloverInstrumentInternalMojo.execute (CloverInstrumentInternalMojo.java:279)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.executeForkedExecutions (MojoExecutor.java:355)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:200)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:64)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: clover.antlr.TokenStreamRecognitionException: expecting '"', found '\n'
    at com.atlassian.clover.instr.java.JavaLexer.nextToken (JavaLexer.java:436)
    at clover.antlr.TokenStreamHiddenTokenFilter.consume (TokenStreamHiddenTokenFilter.java:38)
    at clover.antlr.TokenStreamHiddenTokenFilter.nextToken (TokenStreamHiddenTokenFilter.java:134)
    at com.atlassian.clover.instr.java.CloverTokenStreamFilter.nextToken (CloverTokenStreamFilter.java:164)
    at clover.antlr.TokenBuffer.fill (TokenBuffer.java:69)
    at clover.antlr.TokenBuffer.LA (TokenBuffer.java:80)
    at clover.antlr.LLkParser.LA (LLkParser.java:52)
    at com.atlassian.clover.instr.java.JavaRecognizer.initializer (JavaRecognizer.java:3825)
    at com.atlassian.clover.instr.java.JavaRecognizer.varInitializer (JavaRecognizer.java:3811)
    at com.atlassian.clover.instr.java.JavaRecognizer.variableDeclarator (JavaRecognizer.java:2947)
    at com.atlassian.clover.instr.java.JavaRecognizer.variableDefinitions (JavaRecognizer.java:1435)
    at com.atlassian.clover.instr.java.JavaRecognizer.field (JavaRecognizer.java:2793)
    at com.atlassian.clover.instr.java.JavaRecognizer.classBlock (JavaRecognizer.java:2229)
    at com.atlassian.clover.instr.java.JavaRecognizer.classDefinition (JavaRecognizer.java:1237)
    at com.atlassian.clover.instr.java.JavaRecognizer.typeDefinition2 (JavaRecognizer.java:1167)
    at com.atlassian.clover.instr.java.JavaRecognizer.typeDefinition (JavaRecognizer.java:1071)
    at com.atlassian.clover.instr.java.JavaRecognizer.compilationUnit (JavaRecognizer.java:983)
    at com.atlassian.clover.instr.java.Instrumenter.instrument (Instrumenter.java:211)
    at com.atlassian.clover.instr.java.Instrumenter.instrument (Instrumenter.java:124)
    at com.atlassian.clover.CloverInstr.execute (CloverInstr.java:110)
    at com.atlassian.clover.CloverInstr.mainImpl (CloverInstr.java:86)
    at com.atlassian.maven.plugin.clover.internal.instrumentation.AbstractInstrumenter.instrumentSources (AbstractInstrumenter.java:205)
    at com.atlassian.maven.plugin.clover.internal.instrumentation.AbstractInstrumenter.instrument (AbstractInstrumenter.java:80)
    at com.atlassian.maven.plugin.clover.CloverInstrumentInternalMojo.execute (CloverInstrumentInternalMojo.java:279)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.executeForkedExecutions (MojoExecutor.java:355)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:200)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:64)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

### PR DESCRIPTION
**Pull request by satob**
Closes #141.
- Add support for text blocks

### CODE REVIEW

File: clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java [hunk 1]
```diff
@@ -20,6 +27,7 @@ public class JavaEnvUtils {
             javaVersion = JAVA_8;
             Class.forName("java.lang.StackWalker");
             javaVersion = JAVA_9;
+            javaVersion = SourceVersion.latestSupported().toString().substring("RELEASE_".length());
```
Comment (marek-parfianowicz) [inline clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java:30]: We should have auto-detection based on java runtime, i.e. find some class introduced in a specific Java release.

File: clover-core/src/main/java/com/atlassian/clover/instr/java/java.g [hunk 2]
```diff
@@ -3026,11 +3026,33 @@ CHAR_LITERAL
...
+
+protected STRING_LITERAL_SINGLE_LINE
     : {nc();}   '"' ( ESC | ~( '"' | '\\' | '\n' | '\r') )* '"'
     ;
 
+protected STRING_LITERAL_TEXT_BLOCK
+    : {nc();} '"' '"' '"' ( '\r' | '\n' ) 
+        (   (  BACKSLASH   '"' '"'  '"' ) =>  BACKSLASH   '"' '"' '"'
+          | ( (BACKSLASH)? '"' '"' ~'"' ) => (BACKSLASH)? '"' '"'
+          | ( (BACKSLASH)? '"'     ~'"' ) => (BACKSLASH)? '"'
```
Comment (marek-parfianowicz) [inline clover-core/src/main/java/com/atlassian/clover/instr/java/java.g:3046]: Why do we need these semantic predicates and three variants of the backslash and quotes? Could you explain this a little bit, please?

File: clover-core/src/test/groovy/com/atlassian/clover/JavaSyntax115CompilationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package com.atlassian.clover
+
+import com.atlassian.clover.util.FileUtils
+import org.apache.tools.ant.util.JavaEnvUtils
```
Comment (marek-parfianowicz) [inline clover-core/src/test/groovy/com/atlassian/clover/JavaSyntax115CompilationTest.groovy:4]: This import seems to be unused. You can replace it with a qualified import of `clover..JavaEnvUtils`.

File: clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                \t \r \n \r \t
+                """);
+
+        System.out.println("""
+                \u005c\ \\u005c
+                """);
+
+        System.out.println("""
+                text \
+                concatinated""");
```
Comment (marek-parfianowicz) [inline clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:27]:
```suggestion
                concatenated""");
```

File: clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java (hunk 1 above)
Comment (satob) [reply clover-core/src/main/java/com/atlassian/clover/util/JavaEnvUtils.java]: I have changed the way to get the Java runtime version to use classes/methods introduced in each version.

File: clover-core/src/main/java/com/atlassian/clover/instr/java/java.g (hunk 2 above)
Comment (satob) [reply clover-core/src/main/java/com/atlassian/clover/instr/java/java.g]: I have reviewed these lines in a little more detail and found the first line is unnecessary, so I have deleted the first line.
The second and third line is necessary to handle the code like this (embedded in `Java15TextBlock.java`):
```java
        System.out.println("""
                "" " " ""
                """);
```

Without the second and third lines, you will get an error like this:
```
ERROR: Instrumentation errorInstrumentation error : /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:10:19:expecting '"', found ' '
com.atlassian.clover.api.CloverException: /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:10:19:expecting '"', found ' '
```

Without semantic predicates, you will also get a similar error on the same line:
```
ERROR: Instrumentation errorInstrumentation error : /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:14:19:expecting '"', found ' '
com.atlassian.clover.api.CloverException: /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:14:19:expecting '"', found ' '
```

File: clover-core/src/test/groovy/com/atlassian/clover/JavaSyntax115CompilationTest.groovy (hunk 3 above)
Comment (satob) [reply clover-core/src/test/groovy/com/atlassian/clover/JavaSyntax115CompilationTest.groovy]: Thank you for pointing this out. I have fixed it.

File: clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java (hunk 4 above)
Comment (satob) [reply clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java]: Thank you for pointing this out. I have fixed it.

**Pull request merged**